### PR TITLE
Add FileStore for TSM files

### PR DESCRIPTION
**Pull request by jwilder**
This adds an initial FileStore implementation that manages references to active TSM files on disk.  It provides functionality to load existing TSM files from a directory, add new files, remove existing ones and handle key deletion using tombstone files.

### CODE REVIEW

File: tsdb/engine/tsm1/data_file.go [hunk 1]
```diff
@@ -185,6 +221,33 @@ func (d *directIndex) Entry(key string, t time.Time) *IndexEntry {
 	return nil
 }
 
+func (d *directIndex) Contains(key string) bool {
```
Comment (otoolep) [inline tsdb/engine/tsm1/data_file.go]: Not sure I follow why this is true. The key might be greater lexicographically that then the max key in the index, no?

File: tsdb/engine/tsm1/data_file.go [hunk 2]
```diff
@@ -649,6 +814,36 @@ func (t *tsmReader) ReadAll(key string) ([]Value, error) {
 	return values, nil
 }
 
+func (t *tsmReader) Close() error {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+
+	if c, ok := t.r.(io.Closer); ok {
```
Comment (otoolep) [inline tsdb/engine/tsm1/data_file.go]: Why wouldn't `tsmReader` not support `io.Closer`?

File: tsdb/engine/tsm1/encoding.go [hunk 3]
```diff
@@ -76,14 +76,14 @@ func (a Values) Encode(buf []byte) ([]byte, error) {
 		panic("unable to encode block type")
 	}
 
-	switch a[0].(type) {
```
Comment (otoolep) [inline tsdb/engine/tsm1/encoding.go]: Are these changes about efficiency?

File: tsdb/engine/tsm1/file_store.go [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+		v, err := f.Read(key, t)
+		if err != nil {
+			return nil, err
+		}
+
+		if len(v) > 0 {
+			return v, nil
+		}
+	}
+	return values, nil
```
Comment (jwilder) [inline tsdb/engine/tsm1/file_store.go]: There is no need to allocate here though.

File: tsdb/engine/tsm1/data_file.go (hunk 1 above)
Comment (otoolep) [inline tsdb/engine/tsm1/data_file.go]: Ah, OK -- I see it now. I quickly read `Entries` as a slice, but it's a function call.

File: tsdb/engine/tsm1/file_store.go (hunk 4 above)
Comment (otoolep) [inline tsdb/engine/tsm1/file_store.go]: OK, I see that. I am just wondering if it is necessary to declare the var up above. For example, would this work?

`return []Value(nil), nil`

Does that allocate? Not a big deal, I am just curious.

Comment (otoolep) [inline tsdb/engine/tsm1/file_store.go]: My idea is that it makes the code a little bit clearer.

Comment (jwilder) [inline tsdb/engine/tsm1/file_store.go]: Ah.  Didn't realized I had a `var values []Value` here, but don't actually use it.  I can just return `nil, nil` here actually.

Comment (otoolep) [inline tsdb/engine/tsm1/file_store.go]: Yeah -- that's what I thought, but assumed you had your reasons. :-)

**Comment by otoolep** [comment]
Generally looks good to me, some minor feedback, +1.

**Pull request merged**